File: ios/Jazzify/Defense/Tutorial/DefenseTutorialNotation.swift
import Foundation

struct DefenseTutorialNotationSettings: Equatable, Sendable {
    var notationInstrumentId: String
    var notationOctaveShift: Int
    var clefOverride: NotationInstrumentClef?
    /// Written transposition semitones; nil = use preset.transposition
    var transpositionOverride: Int?
}

enum DefenseTutorialNotation {
    static func resolveClef(_ settings: DefenseTutorialNotationSettings) -> NotationInstrumentClef {
        if let clefOverride = settings.clefOverride {
            return clefOverride
        }
        return NotationInstrumentCatalog.preset(for: settings.notationInstrumentId).clef
    }

    static func resolveVoicingStaff(_ clef: NotationInstrumentClef) -> Int {
        clef.singleStaffNumber ?? 1
    }

    static func resolveDisplayStaves(_ clef: NotationInstrumentClef) -> [Int] {
        if let staff = clef.singleStaffNumber {
            return [staff]
        }
        return [1, 2]
    }

    static func resolveDefaultOctaveShift(_ clef: NotationInstrumentClef) -> Int {
        clef == .bass ? -1 : 0
    }

    static func formatOctaveShiftLabel(_ shift: Int, isEnglishCopy: Bool) -> String {
        let sign = shift > 0 ? "+" : ""
        return isEnglishCopy
            ? "Written octave \(sign)\(shift)"
            : "記譜オクターブ \(sign)\(shift)"
    }

    static func resolveTransposition(_ settings: DefenseTutorialNotationSettings) -> Int {
        if let transpositionOverride = settings.transpositionOverride {
            return transpositionOverride
        }
        return NotationInstrumentCatalog.preset(for: settings.notationInstrumentId).transposition
    }

    static func resolveWrittenOffset(_ settings: DefenseTutorialNotationSettings) -> Int {
        let preset = NotationInstrumentCatalog.preset(for: settings.notationInstrumentId)
        let transposition = resolveTransposition(settings)
        let adjustedPreset: NotationInstrumentPreset
        if transposition == preset.transposition {
            adjustedPreset = preset
        } else {
            adjustedPreset = NotationInstrumentPreset(
                id: preset.id,
                clef: preset.clef,
                transposition: transposition,
                octaveOffset: preset.octaveOffset,
                labelJa: preset.labelJa,
                labelEn: preset.labelEn
            )
        }
        return NotationInstrumentCatalog.writtenSemitoneOffset(
            preset: adjustedPreset,
            userOctaveShift: settings.notationOctaveShift
        )
    }

    private static let writtenKeyByTransposition: [Int: String] = [
        0: "C",
        -2: "B♭",
        -7: "F",
        -9: "E♭",
    ]

    private static func formatWrittenKeyLabel(_ transpositionSemitones: Int) -> String {
        if let key = writtenKeyByTransposition[transpositionSemitones] {
            return "in \(key)"
        }
        let sign = transpositionSemitones > 0 ? "+" : ""
        return "in C\(sign)\(transpositionSemitones)"
    }

    static func formatTutorialNotationLabel(
        _ settings: DefenseTutorialNotationSettings,
        isEnglishCopy: Bool
    ) -> String {
        let clef = resolveClef(settings)
        let clefLabel = NotationInstrumentCatalog.formatClefLabel(
            clef == .grand ? .treble : clef,
            isEnglishCopy: isEnglishCopy
        )
        let transposition = resolveTransposition(settings)
        let keyLabel = formatWrittenKeyLabel(transposition)
        let sep = isEnglishCopy ? " · " : "・"
        return "\(keyLabel)\(sep)\(clefLabel)"
    }

    static func formatTutorialPlaySubtitle(
        concertMidis: [Int],
        isEnglishCopy: Bool
    ) -> String {
        let label = formatTutorialSoundNoteNames(concertMidis: concertMidis)
        return isEnglishCopy ? "Sound: \(label)" : "音：\(label)"
    }

    static func formatTutorialSoundNoteNames(concertMidis: [Int]) -> String {
        concertMidis.map { formatMidiNoteName($0) }.joined(separator: " · ")
    }

    private static func formatMidiNoteName(_ midi: Int) -> String {
        let pitchClass = ((midi % 12) + 12) % 12
        let octave = midi / 12 - 1
        let names = ["C", "C#", "D", "Eb", "E", "F", "F#", "G", "Ab", "A", "Bb", "B"]
        return "\(names[pitchClass])\(octave)"
    }

    static func defaultSettings(
        notationInstrumentId: String,
        notationOctaveShift: Int = 0
    ) -> DefenseTutorialNotationSettings {
        DefenseTutorialNotationSettings(
            notationInstrumentId: notationInstrumentId,
            notationOctaveShift: notationOctaveShift,
            clefOverride: nil,
            transpositionOverride: nil
        )
    }
}

File: src/game/defense/tutorial/defenseTutorialNotation.ts
import {
  formatNotationClefLabel,
  getNotationInstrumentPreset,
  getWrittenSemitoneOffset,
  transposeKeyFifths,
  type NotationInstrumentClef,
  type NotationInstrumentId,
} from '@/utils/notationInstrument';
export interface DefenseTutorialNotationSettings {
  readonly notationInstrumentId: NotationInstrumentId;
  readonly notationOctaveShift: number;
  readonly clefOverride: NotationInstrumentClef | null;
  /** Written transposition semitones; null = use preset.transposition */
  readonly transpositionOverride: number | null;
}

export const resolveTutorialClef = (
  settings: DefenseTutorialNotationSettings,
): NotationInstrumentClef => {
  if (settings.clefOverride) {
    return settings.clefOverride;
  }
  return getNotationInstrumentPreset(settings.notationInstrumentId).clef;
};

export const resolveTutorialTransposition = (
  settings: DefenseTutorialNotationSettings,
): number => {
  if (settings.transpositionOverride !== null) {
    return settings.transpositionOverride;
  }
  return getNotationInstrumentPreset(settings.notationInstrumentId).transposition;
};

const resolveTutorialPresetWithTransposition = (
  settings: DefenseTutorialNotationSettings,
) => {
  const preset = getNotationInstrumentPreset(settings.notationInstrumentId);
  const transposition = resolveTutorialTransposition(settings);
  return transposition === preset.transposition
    ? preset
    : { ...preset, transposition };
};

export const resolveTutorialWrittenOffset = (
  settings: DefenseTutorialNotationSettings,
): number => getWrittenSemitoneOffset(
  resolveTutorialPresetWithTransposition(settings),
  settings.notationOctaveShift,
);

/** Concert key signature for the instrument's sounding do-re-mi (Bb=-2, F=-1, Eb=-3). */
export const resolveTutorialConcertKeyFifths = (
  settings: DefenseTutorialNotationSettings,
): number => transposeKeyFifths(0, resolveTutorialTransposition(settings));

const WRITTEN_KEY_BY_TRANSPOSITION: Readonly<Record<number, string>> = {
  0: 'C',
  [-2]: 'B♭',
  [-7]: 'F',
  [-9]: 'E♭',
};

const formatWrittenKeyLabel = (transpositionSemitones: number): string => {
  const key = WRITTEN_KEY_BY_TRANSPOSITION[transpositionSemitones];
  if (key) {
    return `in ${key}`;
  }
  const sign = transpositionSemitones > 0 ? '+' : '';
  return `in C${sign}${transpositionSemitones}`;
};

/** e.g. "in E♭・ト音記号" */
export const formatTutorialNotationLabel = (
  settings: DefenseTutorialNotationSettings,
  isEnglishCopy: boolean,
): string => {
  const clef = resolveTutorialClef(settings);
  const clefLabel = formatNotationClefLabel(clef === 'grand' ? 'treble' : clef, isEnglishCopy);
  const transposition = resolveTutorialTransposition(settings);
  const keyLabel = formatWrittenKeyLabel(transposition);
  const sep = isEnglishCopy ? ' · ' : '・';
  return `${keyLabel}${sep}${clefLabel}`;
};

export const formatTutorialPlaySubtitle = (
  concertMidis: readonly [number, number, number],
  isEnglishCopy: boolean,
): string => {
  const label = formatTutorialSoundNoteNames(concertMidis);
  return isEnglishCopy ? `Sound: ${label}` : `音：${label}`;
};

export const formatTutorialSoundNoteNames = (
  concertMidis: readonly [number, number, number],
): string => (
  concertMidis.map((midi) => formatMidiNoteName(midi)).join(' · ')
);

const formatMidiNoteName = (midi: number): string => {
  const pitchClass = ((midi % 12) + 12) % 12;
  const octave = Math.floor(midi / 12) - 1;
  const names = ['C', 'C#', 'D', 'Eb', 'E', 'F', 'F#', 'G', 'Ab', 'A', 'Bb', 'B'] as const;
  return `${names[pitchClass]}${octave}`;
};
